src/transport.rs

```diff
@@ -63,112 +81,150 @@ impl Transport {
-            _ => req,
-        };
-        let mut res = embodied.send()?;
-        match res.status {
-            StatusCode::Ok |
-            StatusCode::Created |
-            StatusCode::SwitchingProtocols => Ok(Box::new(res)),
-            StatusCode::NoContent => Ok(
-                Box::new(BufReader::new("".as_bytes())),
-            ),
-            // todo: constantize these
-            StatusCode::BadRequest => {
-                Err(Error::Fault {
-                    code: res.status,
-                    message: get_error_message(&mut res).unwrap_or(
-                        "bad parameter"
-                            .to_owned(),
-                    ),
-                })
+            Transport::Tcp { ref host, .. } => {
+                builder.method(method).uri(&format!("{}{}", host, endpoint))
             }
-            StatusCode::NotFound => {
-                Err(Error::Fault {
-                    code: res.status,
-                    message: get_error_message(&mut res).unwrap_or(
-                        "not found".to_owned(),
-                    ),
-                })
+            Transport::EncryptedTcp { ref host, .. } => {
+                builder.method(method).uri(&format!("{}{}", host, endpoint))
             }
-            StatusCode::NotAcceptable => {
-                Err(Error::Fault {
-                    code: res.status,
-                    message: get_error_message(&mut res).unwrap_or(
-                        "not acceptable"
-                            .to_owned(),
-                    ),
-                })
+            #[cfg(feature = "unix-socket")]
+            Transport::Unix { ref path, .. } => {
+                let uri: hyper::Uri = DomainUri::new(&path, endpoint).into();
+                builder.method(method).uri(&uri.to_string())
             }
-            StatusCode::Conflict => {
-                Err(Error::Fault {
-                    code: res.status,
-                    message: get_error_message(&mut res).unwrap_or(
-                        "conflict found"
-                            .to_owned(),
-                    ),
-                })
+        };
+        let req = req.header(header::HOST, "");
+
+        match body {
+            Some((b, c)) => Ok(req
+                .header(header::CONTENT_TYPE, &c.to_string()[..])
+                .body(b.into())?),
+            _ => Ok(req.body(Body::empty())?),
+        }
+    }
+
+    pub fn stream<B>(
+        &self,
+        method: Method,
+        endpoint: &str,
+        body: Option<(B, Mime)>,
+    ) -> Result<Box<Read>>
+    where
+        B: Into<Body>,
+    {
+        let req = self.build_request(method, endpoint, body)?;
+        let res = self.send_request(req)?;
+
+        match res.status() {
+            StatusCode::OK
+            | StatusCode::CREATED
+            | StatusCode::SWITCHING_PROTOCOLS => {
+                let chunk =
+                    self.runtime().block_on(res.into_body().concat2())?;
```

Comment:
Yes, I suspect you're correct... The whole wrapping of Hyper's async API in a synchronous API is quite clunky unfortunately :( Ideally this would return a `Stream<Vec<u8>>` or similar, and I'm slowly working towards this in a branch but I'm not there yet... I'm not sure how to fix this until then, so I'm open to ideas :)